src/achievements.cpp

```diff
@@ -0,0 +1,116 @@
+/*
+	Copyright (C) 2003 - 2022
+	by David White <dave@whitevine.net>
+	Part of the Battle for Wesnoth Project https://www.wesnoth.org/
+
+	This program is free software; you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation; either version 2 of the License, or
+	(at your option) any later version.
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY.
+
+	See the COPYING file for more details.
+*/
+
+#include "achievements.hpp"
+
+#include "filesystem.hpp"
+#include "game_config.hpp"
+#include "log.hpp"
+#include "preferences/general.hpp"
+#include "serialization/parser.hpp"
+#include "serialization/preprocessor.hpp"
+
+static lg::log_domain log_config("config");
+#define ERR_CONFIG LOG_STREAM(err, log_config)
+
+// TODO: horizontal scrollbar is present despite `horizontal_scrollbar_mode = "never"`
+// TODO: how to make the initial height more than a single achievement? hard code the height? add more achievements?
+// TODO: make checkbox to indicate which achievements have been achieved disabled/untoggleable
+// TODO: GUIWindowDefinitionWML in achievements_dialog.hpp
+// TODO: some way to display in-game when something is achieved?
+
+// TODO: handling translatable strings? needs tstring?
+// TODO: can content_for be translatable? probably not since it's used internally for things - need two separate attributes, one translatable and one not translatable?
```

Comment:
Hmm. It's kinda unclear what it actually refers to. Is it "just" a "unique identifier"? Then it wouldn't be translatable, but you'd still need to somehow map it to a translatable value.

It would be ideal if `content_for` could be an "addon ID", which is basically a directory name, and the game could get the translatable version by looking for that folder and… somehow obtaining it.

Since every `achievements.cfg` is loaded anyway, perhaps this could be achieved by just adding a key in `achievements.cfg` so it looks like this:

```
display_name=_"Tutorial"

[achievement]
    content_for=tutorial
    id=finished
    name=_"Complete the Tutorial"
    description=_"Complete both scenarios of the Tutorial campaign"
[/achievement]
```

The only missing piece here is how to handle elements of mainline. I think namespacing with `/` would do the trick for referring to them, so the above would become something like `content_for=mainline/campaigns/tutorial`, while a `content_for` without any slashes would look in the addons directory.

As for defining the display name for mainline campaigns, that's less clear; your model has a single `achievements.cfg` for all of mainline, but the above approach implies one display name per `achievements.cfg`. Perhaps an `[achievement_group]` tag that contains the `display_name`? An addon could omit it (and use the example above verbatim), but mainline would include several such tags, and addons could use the tag too if they wish. Perhaps they can be nested within each other, so that `/` syntax is actually referring to the hierarchy of `[achievement_group]`s…

Reply:
At the moment it's part of the identifier as well as displayed on the UI, which is the issue. So yeah something like `display_name` would be needed, at minimum.

Reply:
Note that the first form of my proposal would place `display_name` as a toplevel attribute, in case that was unclear.

So the structure I thought up would be something like:

* Toplevel: contains `display_name=`, `[achievement]`, and `[achievement_group]`. The mainline file probably only has `[achievement_group]`.
* `[achievement_group]`: Contains `id=`, `display_name=`, `[achievement]` and `[achievement_group]`.
* `[achievement]`: Contains `content_for=`, `id=`, `icon=`, `name=`, `name_completed=`, `description=`, * `description_completed=`, `hidden=`. Maybe `content_for=` is optional and defaults to a path for the containing achievement group? Or even, it's only allowed in toplevel achievements?

…and that's basically your schema. Does that sound like it works?

The ID used for `[set_achieved]`, `[has_achievement]`, and `content_for=` could be derived from the following format: `[~Addon_Folder_Name / ] (achievement_group_id / )+ achievement_id`, with the achievement ID omitted in `content_for=`.

Reply:
I'm not sure why `display_name` would be a toplevel attribute? If there's an `[achievement_group]`, then it seems like it would make more sense to have `display_name` and `content_for` as required attributes of that tag? So for that (and based on other comments):
```
[achievement_group]
    display_name=
    content_for=
    [achievement]
        name=
        name_completed=
        description=
        description_completed=
        data_uri_icon=
        data_uri_icon_completed=
        hidden=
    [/achievement]
    [achievement]
        name=
        name_completed=
        description=
        description_completed=
        data_uri_icon=
        data_uri_icon_completed=
        hidden=
    [/achievement]
    # etc...
[/achievement_group]
```

Reply:
Well, we could do without it, but it also doesn't seem like we lose anything from making `[achievement_group]` essentially be the root tag, so that the contents of toplevel is the same as the contents of that tag. Then addons that only have a few achievements don't even need to define a group; there's already an implicit group for their addon.

From your code just there, it feels like you have a slightly different concept of `[achievement_group]` than I do. How does `content_for=` work in your model? Is it just an arbitrary ID referenced verbatim directly from `[set_achieved]` or `[has_achievement]`

Reply:
It's meant to be the group ID used by whatever content the achievements are associated to. However, each add-on can have multiple achievement groups (ie: achievements per scenario in a scenario pack) as well as add achievements to entirely unrelated content (ie: a modification that adds more achievements to mainline campaigns).

Reply:
Both my model and yours seem to achieve that… yours looks a little simpler but means we need to be more careful of ID clashes…

Reply:
The translations for all installed addons are already loaded on the title screen, because they're needed for the campaign selection menu. I haven't checked, but it's probable that the WML loading and saving functions are already storing the textdomain in the preferences for you.

Reply:
With the approach Pentarctagon built here, we can't use the campaign title as the display name. An addon is a container that contains one or more root tags such as a campaign, an era, etc. There might not even _be_ a campaign.

I suppose it would be possible to use the addon name though, which would be loaded from _server.pbl or info.cfg.

Oh, but are you trying to say that translatable strings are still serializable? I think that's probably true, but even so, it doesn't seem right to use a translatable string as an identifier.

Reply:
In this case, I believe it would result in (for example) achievements being "lost" if the player changes the language.

Reply:
I'm not confident that that's true, but even if it's not true, it doesn't seem like a good idea.

Reply:
I just meant the display_name would probably be serializable without breaking the translation.

Reply:
@CelticMinstrel What I mean is that if a translatable value were used as an identifier as well then if:
* Someone completed an achievement while using the French translation.
* Closed Wesnoth.
* Started Wesnoth and changed the language to German.
Then they wouldn't have that achievement marked as complete anymore since it would be comparing the French string (from the preferences file) to the German string (current language).

@stevecotton Thanks, good to know.

Reply:
I got what you meant, and I suspect that wouldn't happen, but even then, I don't think it's a good idea to do things that way.

Reply:
This has been addressed/completed.